### PR TITLE
BUG: fix gcd inf

### PR DESCRIPTION
**Pull request by guan404ming**
- fix `gcd()` hangs when passing an array of object dtype containing infinity
- add some tests for the modification
- ref @ngoldbaum  https://github.com/numpy/numpy/issues/26726#issuecomment-2218305985

### CODE REVIEW

File: numpy/_core/_internal.py [hunk 1]
```diff
@@ -860,6 +861,9 @@ def _prod(a):
 
 def _gcd(a, b):
     """Calculate the greatest common divisor of a and b"""
+    if not (math.isfinite(a) and math.isfinite(b)):
+        raise ValueError('Can only find greatest common demoninator of a '
```
Comment (eendebakpt) [inline numpy/_core/_internal.py:865]:
```suggestion
        raise ValueError('Can only find greatest common demoninator of '
```

File: numpy/_core/_internal.py [hunk 2]
```diff
@@ -860,6 +861,9 @@ def _prod(a):
...
+                            f'finite argument, found "{a}" and "{b}"')
```
Comment (eendebakpt) [inline numpy/_core/_internal.py:866]:
```suggestion
                            f'finite arguments, found "{a}" and "{b}"')
```

File: numpy/_core/tests/test_umath.py [hunk 3]
```diff
@@ -4138,6 +4138,11 @@ def test_huge_integers(self):
         assert_equal(np.gcd(a, b), [2**100,               2**50 * 3**5])
         assert_equal(np.lcm(a, b), [2**100 * 3**5 * 5**7, 2**100 * 3**10])
 
+    def test_inf(self):
```
Comment (eendebakpt) [inline numpy/_core/tests/test_umath.py:4141]: Could you also add test cases for nan and a direct `gcd(4, float(`inf`))`?

**Comment by eendebakpt** [comment]
@guan404ming Looks good to me, just two minor suggestions.

Since `_gcd` is used in the path for object dtype the performance impact of the additional check seems ok.

**Comment by guan404ming** [comment]
I have updated the branch based on your suggestions, but I'm not sure if everything was right. If there's anything else I need to fix, please let me know~ Thanks!!

File: numpy/_core/_internal.py [hunk 4]
```diff
@@ -860,6 +861,9 @@ def _prod(a):
 
 def _gcd(a, b):
     """Calculate the greatest common divisor of a and b"""
+    if not (math.isfinite(a) and math.isfinite(b)):
+        raise ValueError('Can only find greatest common demoninator of '
```
Comment (eendebakpt) [inline numpy/_core/_internal.py:865]:
```suggestion
        raise ValueError('Can only find greatest common divisor of '
```

File: numpy/_core/tests/test_umath.py [hunk 5]
```diff
@@ -4138,6 +4138,14 @@ def test_huge_integers(self):
...
```
Comment (eendebakpt) [inline numpy/_core/tests/test_umath.py:4141]:
```suggestion
    def test_inf_and_nan(self):
```

**Review by eendebakpt — APPROVED** [review]
@guan404ming Thanks for addressing the comments.

**Comment by guan404ming** [comment]
@eendebakpt Thanks lots for your assistence~

File: numpy/_core/_internal.py [hunk 6]
```diff
@@ -860,6 +861,9 @@ def _prod(a):
 
 def _gcd(a, b):
     """Calculate the greatest common divisor of a and b"""
+    if not (math.isfinite(a) and math.isfinite(b)):
+        raise ValueError('Can only find greatest common divisor of '
+                            f'finite arguments, found "{a}" and "{b}"')
```
Comment (ngoldbaum) [inline numpy/_core/_internal.py:866]:
```suggestion
        raise ValueError('Can only find greatest common divisor of '
                         f'finite arguments, found "{a}" and "{b}"')
```

Alignment is off here

**Pull request merged**